lib/dry/schema/primitive_inferrer.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require 'dry/core/cache'
+
+module Dry
+  module Schema
+    # PrimitiveInferrer is used internally by `Macros::Filled`
+    # for inferring a list of possible primitives that a given
+    # type can handle.
+    #
+    # @api private
+    class PrimitiveInferrer
+      extend Dry::Core::Cache
+
+      # Compiler reduces type AST into a list of primitives
+      #
+      # @api private
+      class Compiler
+        # @api private
+        def visit(node)
+          meth, rest = node
+          public_send(:"visit_#{meth}", rest)
+        end
+
+        # @api private
+        def visit_nominal(node)
+          type, _ = node
+          [type]
+        end
+
+        # @api private
+        def visit_hash(_)
+          Hash
```

Comment:
`visit_*` methods should always return an array, otherwise, `Filled#primitives` may return a class instead of an array of classes